File: backend/src/services/agent-runner/office_helpers.py
"""Stdlib-only Office helpers the agent can import from /workspace/tmp/office_helpers.py.

No python-pptx required. Used for slide clone / text inspect so any request
(white, pink, hex, thanks slide, a comma) can be verified or applied.
"""
from __future__ import annotations

import io
import re
import zipfile


def _slide_nums(names):
    out = []
    for n in names:
        m = re.match(r"ppt/slides/slide(\d+)\.xml$", n)
        if m:
            out.append(int(m.group(1)))
    return sorted(out)


def list_slide_texts(path):
    texts = []
    with zipfile.ZipFile(path) as z:
        for n in _slide_nums(z.namelist()):
            xml = z.read(f"ppt/slides/slide{n}.xml").decode("utf-8", "ignore")
            texts.append(" ".join(re.findall(r"<a:t[^>]*>([^<]*)</a:t>", xml)))
    return texts


def xml_has_hex(path, hex_color):
    needle = hex_color.replace("#", "").upper()
    with zipfile.ZipFile(path) as z:
        for name in z.namelist():
            if not name.endswith(".xml"):
                continue
            if needle in z.read(name).decode("utf-8", "ignore").upper():
                return True
    return False


def append_text_slide(src, title, dst=None):
    """Clone the last slide, put `title` in the first text run, write dst."""
    dst = dst or src
    with zipfile.ZipFile(src, "r") as zin:
        names = zin.namelist()
        nums = _slide_nums(names)
        if not nums:
            raise RuntimeError("pptx has no slides")
        last_n = nums[-1]
        new_n = last_n + 1
        last_xml = zin.read(f"ppt/slides/slide{last_n}.xml").decode("utf-8")
        new_xml, nsub = re.subn(r"(<a:t[^>]*>)([^<]*)(</a:t>)", rf"\1{title}\3", last_xml, count=1)
        if nsub == 0:
            raise RuntimeError("could not find a text run to rewrite")

        ct = zin.read("[Content_Types].xml").decode("utf-8")
        if f"slide{new_n}.xml" not in ct:
            ct = ct.replace(
                "</Types>",
                f'<Override PartName="/ppt/slides/slide{new_n}.xml" '
                f'ContentType="application/vnd.openxmlformats-officedocument.presentationml.slide+xml"/>'
                f"</Types>",
            )

        rels = zin.read("ppt/_rels/presentation.xml.rels").decode("utf-8")
        rids = [int(x) for x in re.findall(r'Id="rId(\d+)"', rels)]
        new_rid = (max(rids) if rids else 10) + 1
        if f"slide{new_n}.xml" not in rels:
            rels = rels.replace(
                "</Relationships>",
                f'<Relationship Id="rId{new_rid}" '
                f'Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/slide" '
                f'Target="slides/slide{new_n}.xml"/></Relationships>',
            )

        pres = zin.read("ppt/presentation.xml").decode("utf-8")
        sids = [int(x) for x in re.findall(r'<p:sldId[^>]*id="(\d+)"', pres)]
        new_sid = (max(sids) if sids else 255) + 1
        if f'rId{new_rid}"' not in pres:
            if "</p:sldIdLst>" not in pres:
                raise RuntimeError("presentation.xml has no sldIdLst")
            pres = pres.replace(
                "</p:sldIdLst>",
                f'<p:sldId id="{new_sid}" r:id="rId{new_rid}"/></p:sldIdLst>',
            )

        buf = io.BytesIO()
        with zipfile.ZipFile(buf, "w") as zout:
            for name in names:
                if name == f"ppt/slides/slide{new_n}.xml":
                    continue
                data = zin.read(name)
                if name == "[Content_Types].xml":
                    data = ct.encode("utf-8")
                elif name == "ppt/_rels/presentation.xml.rels":
                    data = rels.encode("utf-8")
                elif name == "ppt/presentation.xml":
                    data = pres.encode("utf-8")
                zout.writestr(name, data)
            zout.writestr(f"ppt/slides/slide{new_n}.xml", new_xml)
            rels_old = f"ppt/slides/_rels/slide{last_n}.xml.rels"
            if rels_old in names:
                zout.writestr(
                    f"ppt/slides/_rels/slide{new_n}.xml.rels",
                    zin.read(rels_old),
                )
        open(dst, "wb").write(buf.getvalue())
    return dst

File: backend/src/services/agent-runner/office-helpers.js
'use strict';

/**
 * Stdlib-ish OOXML helpers used by AgentRunner's optional high-level path.
 * The model can still do the same job with execute_python; this exists so
 * "agrega una lámina de gracias" is reliable without python-pptx.
 */

const PizZip = require('pizzip');

function xmlEscape(value) {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function slideNumbers(zip) {
  return Object.keys(zip.files)
    .map((n) => {
      const m = n.match(/^ppt\/slides\/slide(\d+)\.xml$/);
      return m ? Number(m[1]) : null;
    })
    .filter((n) => Number.isFinite(n))
    .sort((a, b) => a - b);
}

function appendTextSlide({ buffer, title = 'Gracias' } = {}) {
  if (!Buffer.isBuffer(buffer) || !buffer.length) {
    throw new Error('appendTextSlide: buffer is required');
  }
  const zip = new PizZip(buffer);
  const nums = slideNumbers(zip);
  if (!nums.length) throw new Error('appendTextSlide: pptx has no slides');
  const lastN = nums[nums.length - 1];
  const newN = lastN + 1;
  const lastXml = zip.file(`ppt/slides/slide${lastN}.xml`)?.asText();
  if (!lastXml) throw new Error('appendTextSlide: last slide missing');

  let newXml = lastXml.replace(/(<a:t[^>]*>)([^<]*)(<\/a:t>)/, `$1${xmlEscape(title)}$3`);
  if (newXml === lastXml) {
    throw new Error('appendTextSlide: no text run to rewrite');
  }
  zip.file(`ppt/slides/slide${newN}.xml`, newXml);

  const lastRels = zip.file(`ppt/slides/_rels/slide${lastN}.xml.rels`);
  if (lastRels) {
    zip.file(`ppt/slides/_rels/slide${newN}.xml.rels`, lastRels.asText());
  }

  let ct = zip.file('[Content_Types].xml')?.asText() || '';
  if (!ct.includes(`slide${newN}.xml`)) {
    ct = ct.replace(
      '</Types>',
      `<Override PartName="/ppt/slides/slide${newN}.xml" ContentType="application/vnd.openxmlformats-officedocument.presentationml.slide+xml"/></Types>`,
    );
    zip.file('[Content_Types].xml', ct);
  }

  let rels = zip.file('ppt/_rels/presentation.xml.rels')?.asText() || '';
  const rids = [...rels.matchAll(/Id="rId(\d+)"/g)].map((m) => Number(m[1]));
  const newRid = (rids.length ? Math.max(...rids) : 10) + 1;
  if (!rels.includes(`slide${newN}.xml`)) {
    rels = rels.replace(
      '</Relationships>',
      `<Relationship Id="rId${newRid}" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/slide" Target="slides/slide${newN}.xml"/></Relationships>`,
    );
    zip.file('ppt/_rels/presentation.xml.rels', rels);
  }

  let pres = zip.file('ppt/presentation.xml')?.asText() || '';
  const sids = [...pres.matchAll(/<p:sldId[^>]*id="(\d+)"/g)].map((m) => Number(m[1]));
  const newSid = (sids.length ? Math.max(...sids) : 255) + 1;
  if (!pres.includes(`rId${newRid}"`)) {
    if (!pres.includes('</p:sldIdLst>')) {
      throw new Error('appendTextSlide: presentation.xml has no sldIdLst');
    }
    pres = pres.replace(
      '</p:sldIdLst>',
      `<p:sldId id="${newSid}" r:id="rId${newRid}"/></p:sldIdLst>`,
    );
    zip.file('ppt/presentation.xml', pres);
  }

  return {
    buffer: zip.generate({ type: 'nodebuffer' }),
    slideNumber: newN,
    title: String(title),
  };
}

module.exports = { appendTextSlide, slideNumbers };
